src/main/scala/cognite/spark/v1/NumericDataPointsRdd.scala

```diff
@@ -422,54 +434,81 @@ case class NumericDataPointsRdd(
             id,
             dataPointsRange.start,
             dataPointsRange.end)
+          .map { allDataPoints =>
+            increaseReadMetrics(allDataPoints.length)
+            allDataPoints.map(p => toRowDataPointsRange(dataPointsRange.id, p))
+          }
+    }
+    Stream.evalUnChunk(points.map(Chunk.seq))
+  }
+
+  @inline
+  private def toRowDataPointsRange(id: Either[Long, String], dataPoint: SdkDataPoint): Row = {
+    val array = new Array[Any](rowIndices.length)
+    var i = 0
+    for (f <- rowIndices) {
+      f match {
+        case 0 => array(i) = id.left.toOption
+        case 1 => array(i) = id.right.toOption
+        case 2 => array(i) = java.sql.Timestamp.from(dataPoint.timestamp)
+        case 3 => array(i) = dataPoint.value
+        case 4 | 5 => array(i) = None
+      }
+      i += 1
+    }
```

Comment:
```suggestion
    for ((f, i) <- rowIndices.zipWithIndex) {
      f match {
        case 0 => array(i) = id.left.toOption
        case 1 => array(i) = id.right.toOption
        case 2 => array(i) = java.sql.Timestamp.from(dataPoint.timestamp)
        case 3 => array(i) = dataPoint.value
        case 4 | 5 => array(i) = None
      }
    }
```

Reply:
As far as I can that leads to less efficient code (which allocates tuples, and ). It's ugly, but it's meant to be very fast and efficient since it gets called hundreds of thousands of times (once for every data point).

From "Decompile Scala to Java" in IntelliJ, the first version is mine:
```java
   private Row toRowDataPointsRange(final Either id, final DataPoint dataPoint) {
      Object[] array = new Object[this.rowIndices().length];
      IntRef i = IntRef.create(0);
      (new ofInt(scala.Predef..MODULE$.intArrayOps(this.rowIndices()))).foreach((f) -> {
         switch(f) {
         case 0:
            array[i.elem] = id.left().toOption();
            break;
         case 1:
            array[i.elem] = id.right().toOption();
            break;
         case 2:
            array[i.elem] = Timestamp.from(dataPoint.timestamp());
            break;
         case 3:
            array[i.elem] = BoxesRunTime.boxToDouble(dataPoint.value());
            break;
         case 4:
         case 5:
            array[i.elem] = scala.None..MODULE$;
            break;
         default:
            throw new MatchError(BoxesRunTime.boxToInteger(f));
         }

         ++i.elem;
      });
      return new GenericRow(array);
   }

   private Row toRowDataPointsRange(final Either id, final DataPoint dataPoint) {
      Object[] array = new Object[this.rowIndices().length];
      (new ofRef(scala.Predef..MODULE$.refArrayOps((Object[])(new ofInt(scala.Predef..MODULE$.intArrayOps(this.rowIndices()))).zipWithIndex(scala.Array..MODULE$.canBuildFrom(scala.reflect.ClassTag..MODULE$.apply(Tuple2.class)))))).withFilter((check$ifrefutable$3) -> {
         return BoxesRunTime.boxToBoolean($anonfun$toRowDataPointsRange$1(check$ifrefutable$3));
      }).foreach((x$23) -> {
         $anonfun$toRowDataPointsRange$2(array, id, dataPoint, x$23);
         return BoxedUnit.UNIT;
      });
      return new GenericRow(array);
   }

   // $FF: synthetic method
   public static final void $anonfun$toRowDataPointsRange$2(final Object[] array$1, final Either id$5, final DataPoint dataPoint$1, final Tuple2 x$23) {
      if (x$23 != null) {
         int f = x$23._1$mcI$sp();
         int i = x$23._2$mcI$sp();
         BoxedUnit var10000;
         switch(f) {
         case 0:
            array$1[i] = id$5.left().toOption();
            var10000 = BoxedUnit.UNIT;
            break;
         case 1:
            array$1[i] = id$5.right().toOption();
            var10000 = BoxedUnit.UNIT;
            break;
         case 2:
            array$1[i] = Timestamp.from(dataPoint$1.timestamp());
            var10000 = BoxedUnit.UNIT;
            break;
         case 3:
            array$1[i] = BoxesRunTime.boxToDouble(dataPoint$1.value());
            var10000 = BoxedUnit.UNIT;
            break;
         case 4:
         case 5:
            array$1[i] = scala.None..MODULE$;
            var10000 = BoxedUnit.UNIT;
            break;
         default:
            throw new MatchError(BoxesRunTime.boxToInteger(f));
         }

      } else {
         throw new MatchError(x$23);
      }
   }
```

It's always hard to predict what the JVM will turn this into once it's JITted, but I'd prefer the stupid, simple, ugly and definitely fast version in this case. ;)